src/credentials/DeviceAttestationVendorReserved.h

```diff
@@ -0,0 +1,219 @@
+
+    // read TLV until first profile tag
+    CHIP_ERROR PrepareToReadVendorReservedElements(const ByteSpan & attestationElements, size_t count)
+    {
+        mIsInitialized         = false;
+        mNumVendorReservedData = count;
+        mAttestationData       = attestationElements;
+
+        tlvReader.Init(mAttestationData);
+        ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+        ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+
+        // position to first ProfileTag
+        while (true)
+        {
+            ReturnErrorOnFailure(tlvReader.Next());
+            if (!TLV::IsProfileTag(tlvReader.GetTag()))
+                break;
+        }
+        // positioned to first context tag (vendor reserved data)
+        mIsInitialized = true;
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Next() { return tlvReader.Next(); }
+
+    size_t GetNumberOfElements() { return mNumVendorReservedData; }
+
+    /**
+     *  @brief Return next VendorReserved element. PrepareToReadVendorReservedElements must be called first.
+     *
+     *  @param[out] element  Next vendor Reserved element
+     *
+     *  @returns   CHIP_NO_ERROR on success
+     *             CHIP_ERROR_INCORRECT_STATE if PrepareToReadVendorReservedElements hasn't been called first
+     *             CHIP_END_OF_TLV if not further entries are present
+     */
+    CHIP_ERROR GetNextVendorReservedElement(struct VendorReservedElement & element)
+    {
+        CHIP_ERROR err;
+
+        VerifyOrReturnError(mIsInitialized, CHIP_ERROR_INCORRECT_STATE);
+
+        while ((err = Next()) == CHIP_NO_ERROR)
+        {
+            uint64_t tag = tlvReader.GetTag();
+            if (!TLV::IsProfileTag(tag))
+            {
+                continue;
+            }
+            // tag is profile tag
+            element.vendorId   = TLV::VendorIdFromTag(tag);
+            element.profileNum = TLV::ProfileNumFromTag(tag);
+            element.tagNum     = TLV::TagNumFromTag(tag);
+
+            return tlvReader.GetByteView(element.vendorReservedData);
+        }
+
+        return err;
+    }
+
+private:
+    size_t mNumVendorReservedData; // number of VendorReserved entries (could be 0)
+    ByteSpan mAttestationData;
+    bool mIsInitialized = false;
+    TLV::ContiguousBufferTLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+};
+
+class DeviceAttestationVendorReservedConstructor
+{
+public:
+    DeviceAttestationVendorReservedConstructor(struct VendorReservedElement * array, size_t size) : mElements(array), mMaxSize(size)
+    {}
+
+    typedef const struct VendorReservedElement * const_iterator;
+
+    const_iterator Next()
+    {
+        VerifyOrReturnError(mCurrentIndex < mNumEntriesUsed, nullptr);
+        return &mElements[mCurrentIndex++];
+    }
+
+    const_iterator cbegin()
+    {
+        do_sorting(); // when beginning to iterator, make a linked list and return the head element
```

Comment:
There is no linked list. This sorts the array in-place. Please update the comment

Reply:
Addressed in https://github.com/project-chip/connectedhomeip/pull/10546/commits/e0bfd9376e680b8fcd5323e4eb85995dbd0fab3c